### PR TITLE
Testing GMSO XML handling and Conversions, Forcefield-utilities backend

### PR DESCRIPTION
**Pull request by CalCraven**
This PR aims to show exactly which sets of XML files we can write for various forcefields, unit styles, and molecule types. These forcefields will be placed into a library of forcefields and tested using our GMSO ForceField reader, writer, and an `__eq__` method. We can also use this set of ForceFields as a starting place for testing for intra-convertibility.
 
Another big change promoted in this PR is the deprecation of the GMSO Forcefield loader and replacement with the forcefield-utilities loader.

To do list:
* [x] forcefield-utilities backend
* [x] deprecate gmso loader
* [x] test for reading in forcefields
* [x] test for writing out forcefields
* [x] test for forcefield equivalency, inter-convertibility, inter-compatibility
* [x] set of forcefields by unit_styles
* [x] set of forcefields by forcefield type
* [x] set of forcefields by molecule
* [ ] modifications to `test_forcefield.py` to check new forcefields

### CODE REVIEW

**Comment by CalCraven** [comment]
Lists of styles to be supported.
* [x] [Charmm36](https://github.com/andersx/charmm36/blob/master/test/parameters/param_all36_eef1sb.prm)
* [x] tip3p 
* [x] tip4p_2005
* [x] tip4p_ewald
* [ ] [ff19SB](https://doi.org/10.1021/acs.jctc.9b00591) https://github.com/csimmerling/ff19SB_201907, https://github.com/intbio/gromacs_ff/tree/master/amber14sb_OL15.ff
* [ ] OL15
* [ ] OL3
* [ ] gaff
* [ ] lipids21
* [ ] DRUDE
* [ ] AMOEBA
* [ ] Martini
* [ ] UNRES

**Comment by bc118** [comment]
This is now passing all the test with mosdef-gomc when pulling the foyer, mbuild, and forcefield-utilities "mains" so it is approved on this basis of testing.

File: gmso/utils/misc.py
Comment (daico007) [inline gmso/utils/misc.py:18]:
```suggestion
    for unit1, unit2 in zip(units1, units2):
        try:
            u.testing.assert_allclose_units(unit1, unit2, rtol=1e-5, atol=1e-8)
        except AssertionError:
            return False
    return True
```

File: gmso/utils/compatibility.py
Comment (daico007) [inline gmso/utils/compatibility.py:53]: Forgotten print statement

File: gmso/utils/files/gmso_xmls/test_ffstyles/oplsaa-ethane_foyer.xml
Comment (daico007) [inline gmso/utils/files/gmso_xmls/test_ffstyles/oplsaa-ethane_foyer.xml:1]: I think we can rename this file to be just `oplsaa_ethane.xml`

File: gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt
Comment (daico007) [inline gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt:1]: Think this file get wrong extension

**Review by daico007 — CHANGES_REQUESTED** [review]
Looks good overall, only some small changes needed before merge.

File: gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt
Comment (bc118) [reply gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt:1]: I also think @daico007 is right.

**Review by bc118 — CHANGES_REQUESTED** [review]
This looks pretty good, just look at my comments and see if they make sense.  

This all passed for all 4 mosdef-gomc test  files when I pulled the main ff-utils foyer and mbuild and the Cal branch gmso --branch 690-testing-xml-writer.  Therefore, I believe I tested it independently under a wide range of test conditions.

File: gmso/utils/misc.py
Comment (CalCraven) [reply gmso/utils/misc.py:18]: The testing.assert_allclose_units function raises an Assertion error itself if unit1 and unit2 are not within tolerance. So that should fail the try statement and go to the except and return False, which will then fail the assertion I laid out in the main __eq__ check. There's probably a better way around this, I'm open to thoughts. I just didn't want this check to raise an assertion error everytime you check equality, it should be return True or False.

Comment (CalCraven) [reply gmso/utils/misc.py:18]: This works just as well I believe, although I think it still runs into the fundamental issue that looking for an AssertionError isn't exactly what we want to do here.

File: gmso/utils/files/gmso_xmls/test_ffstyles/oplsaa-ethane_foyer.xml
Comment (CalCraven) [reply gmso/utils/files/gmso_xmls/test_ffstyles/oplsaa-ethane_foyer.xml:1]: I agree the name isn't great here, it was copied from the gmso testing suite. I changed the name to oplsaa_from_foyer.xml to better reflect it's origin and what we're testing for with it. There's already an alkanes.xml that should be the same as the oplsaa_ethane.xml example we would want to cover.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py:103]: I can add it in. We haven't written ffutils docs yet, but typically I import forcefield-utilities as ffutils so I wanted to cover that case mostly.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py:562]: I think we can actually leave it for now until we want to decide what to do with it. I'll mark it as a TODO. I just listed out all of the attribute of a forcefield here to check for equality, but this was the last one that won't look the same unless we make changes to the main default reader in GMSO. But if we deprecate it out, then we won't have to worry about the equality checks being incorrect.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py:597]: Added in a NotImplemented Error for it.

File: gmso/tests/test_atom_type.py
Comment (CalCraven) [reply gmso/tests/test_atom_type.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96 with a new private function _xnl_from_gmso. Great suggestion.

File: gmso/utils/compatibility.py
Comment (CalCraven) [reply gmso/utils/compatibility.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py:104]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96.

File: gmso/core/forcefield.py
Comment (CalCraven) [reply gmso/core/forcefield.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96.

File: gmso/parameterization/foyer_utils.py
Comment (CalCraven) [reply gmso/parameterization/foyer_utils.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96.

File: gmso/tests/test_forcefield.py
Comment (CalCraven) [reply gmso/tests/test_forcefield.py]: Fixed in commit 93d813f82a66dc81a825042b283a26b3b4890c96. I had changed them to see if the tests passed for these xmls because they didn't for the old versions of the charmm and buck forcefields, but then once I got the okay to manually change their formatting to match the updated formats, they pass the tests now fine. So I changed it back to test the charmm and buck ffs again.

File: gmso/core/forcefield.py
Comment (bc118) [reply gmso/core/forcefield.py:563]: That's fine with me

File: gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt
Comment (bc118) [reply gmso/utils/files/gmso_xmls/test_ffstyles/ff14SB.txt:1]: OK

File: gmso/core/forcefield.py
Comment (bc118) [reply gmso/core/forcefield.py]: Fine with me

File: gmso/tests/test_forcefield.py
Comment (bc118) [reply gmso/tests/test_forcefield.py]: OK

**Review by daico007 — APPROVED** [review]
Everything LGTM! I believe the failing tests are depending on this PR is ffutils https://github.com/mosdef-hub/forcefield-utilities/pull/28, so probably need to wait for that PR is in first for this PR to be merged

**Comment by daico007** [comment]
Bleeding test appears to pass after this [PR](https://github.com/mosdef-hub/forcefield-utilities/pull/28) is merged. Will merge this first thing in the morning (if there is no extra comment/objection)

**Pull request merged**
